src/v/cluster/controller_backend.cc

```diff
@@ -885,522 +690,491 @@ void controller_backend::housekeeping() {
+    // partition should exist on both new nodes (those not present in the
+    // original assignment) and the original nodes for the whole duration of the
+    // update. On original nodes, if there is a cross-shard move, the shard is
+    // determined by the end state of the move.
+    //
+    // Example: (x/y means node/shard) suppose there is an update
+    // {1/0, 2/0, 3/0} -> {2/1, 3/0, 4/0} in progress. Then the partition object
+    // must be present only on 1/0, 2/1, 3/0 and 4/0. If the update is then
+    // cancelled, the partition object must be present only on 1/0, 2/0, 3/0,
+    // 4/0 (note that the desired shard for node 2 changed).
+    //
+    // Note also that the partition on new nodes is always created, even if the
+    // update is cancelled. This is due to a possibility of "revert_cancel"
+    // scenario (i.e. when we cancel a move, but the raft layer has already
+    // completed it, in this case we end up with the "updated" replica set, not
+    // the original one).
+
+    auto as_it = topic_it->second.get_assignments().find(ntp.tp.partition);
+    vassert(
+      as_it != topic_it->second.get_assignments().end(),
+      "[{}] expected assignment",
+      ntp);
+    const auto& assignment = *as_it;
+
+    const replicas_t* orig_replicas = &assignment.replicas;
+    const replicas_t* updated_replicas = nullptr;
+    model::revision_id last_cmd_revision
+      = p_it->second.last_update_finished_revision;
+
+    auto update_it = _topics.local().updates_in_progress().find(ntp);
+    if (update_it != _topics.local().updates_in_progress().end()) {
+        vlog(clusterlog.trace, "[{}] update: {}", ntp, update_it->second);
+        orig_replicas = &update_it->second.get_previous_replicas();
+        updated_replicas = &update_it->second.get_target_replicas();
+        last_cmd_revision = update_it->second.get_last_cmd_revision();
+    }
+
+    std::optional<ss::shard_id> orig_shard_on_cur_node = find_shard_on_node(
+      *orig_replicas, _self);
+    std::optional<ss::shard_id> updated_shard_on_cur_node
+      = updated_replicas ? find_shard_on_node(*updated_replicas, _self)
+                         : std::nullopt;
+
+    // Has value if the partition is expected to exist on this node.
+    std::optional<model::revision_id> expected_log_rev;
+    if (orig_shard_on_cur_node) {
+        expected_log_rev = p_it->second.replicas_revisions.at(_self);
+    } else if (updated_shard_on_cur_node) {
+        expected_log_rev = update_it->second.get_update_revision();
+    }
+
+    bool expected_on_cur_shard = false;
+    if (orig_shard_on_cur_node) {
+        if (updated_shard_on_cur_node) {
+            // update on this node is a cross-core move
+            expected_on_cur_shard = has_local_replicas(
+              _self, assignment.replicas);
+        } else {
+            expected_on_cur_shard
+              = (orig_shard_on_cur_node == ss::this_shard_id());
```

Comment:
nit: i think this would benefit from a comment describing a particular case handled in each `if/else` branch f.e. `// partition is moving to current shard from other node`

Reply:
will add